students/mkdir01/lesson09/mailroomOO.py

```diff
@@ -0,0 +1,269 @@
+            else:
+                raise KeyError(data)
+
+    def __setitem__(self, num, name):
+        pass
+
+    def add_donor(self, name):
+        if type(name) == Individual_Donor:
+            self._donor_obj_list.append(name)
+        else:
+            raise TypeError('object must be type \'Individual_Donor\'')
+
+    def add_donation(self, name, amt, day=str(date.today())):
+        if name in self.donor_list:
+            self[name].add_dollars(amt, day)
+        else:
+            self.add_donor(Individual_Donor(name, amt, day))
+
+    def save_donor_list(self, filename):
+        temp_dict = {}
+        for obj in self.donor_obj_list:
+            temp_dict[obj.name] = obj.donations
+        with open(filename, 'w') as outfile:
+            outfile.write(str(temp_dict))
+
+    def load_donor_list(self, filename):
+        temp_str = open(filename).read()
+        temp_data = self.parser(temp_str)
+        for name in temp_data:
+            for day in temp_data[name]:
+                for amt in temp_data[name][day]:
+                    self.add_donation(name, amt, day=day)
+
+    def parser(self, str):
+        return ast.literal_eval(str)
+###
+
+
+### User interactions ###
+# mode 1 - add a donation and send a thank you
+def input_name(obj):
+    name = input("Please enter a full name (or type \'list\', or hit \'return\' to go back)> ")
+    if name == "list":
+            print('\n')
+            print('\n'.join(obj.donor_list))
+            print('\n')
+    elif name == "":
+        return None
+    else:
+        return name
+
+
+def input_dollars(name, obj):
+    dollars = input("Please enter a donation amount (ex: 500.05)> ")
+    try:
+        float(dollars)
+        if float(dollars) > 0:
+            obj.add_donation(name, float(dollars))
+            return float(dollars)
+        else:
+            return None
+    except ValueError:
+        raise ValueError("Donation amounts must be numbers")
+
+
+def thank_you(name, dollars):
+    print(f"Dear {name},\n\nThank you for your donation of {dollars:.2f}.\n\nLove,\nThe Charity Org")
+
+
+def add_donation_to_list(obj):
+    while True:
+        name = input_name(obj)
+        if not name:
+            break
+        dollars = input_dollars(name, obj)
+        if name and dollars:
+            thank_you(name, dollars)
+
+
+# mode 2 - "create a report"
+def sort_by_dollars(donor_lists):
+    try:
+        donor_lists.sort(key=lambda k: k[1], reverse=True)
+    except IndexError:
+        print('This list can not be sorted.')
+        print(donor_lists)
+
+
+def create_donor_list(obj):
+    # create [[donor name, total donations, # gifts, avg gift], [donor 2 info]]
+    temp_list = []
+    for i, name in enumerate(obj.donor_list):
+        temp_list.append([name, obj[name].report_total_donations(), obj[name].report_total_num_gifts(), (obj[name].report_total_donations() / obj[name].report_total_num_gifts())])
+    sort_by_dollars(temp_list)
+    return temp_list
+
+
```

Comment:
You might consider moving some of these module level functions into the `Group_Donors` class for better encapsulation